File: db4on/reference/db4o-reference-chapters/src/com/db4odoc/f1/clientserver/TransactionExample.cs
using System;
using System.IO;
using com.db4o;
using com.db4odoc.f1;

namespace com.db4odoc.f1.clientserver
{
    public class TransactionExample 
    {
		public readonly static string YapFileName = "formula1.yap";

        public static void Main(string[] args)
        {
            File.Delete(YapFileName);
            ObjectContainer db=Db4o.OpenFile(YapFileName);
            try
            {
                StoreCarCommit(db);
                db.Close();
                db = Db4o.OpenFile(YapFileName);
                ListAllCars(db);
                StoreCarRollback(db);
                db.Close();
                db = Db4o.OpenFile(YapFileName);
                ListAllCars(db);
                CarSnapshotRollback(db);
                CarSnapshotRollbackRefresh(db);
            }
            finally
            {
                db.Close();
            }
        }
		// end Main
        
        public static void StoreCarCommit(ObjectContainer db)
        {
            Pilot pilot = new Pilot("Rubens Barrichello", 99);
            Car car = new Car("BMW");
            car.Pilot = pilot;
            db.Set(car);
            db.Commit();
        }
		// end StoreCarCommit
    
        public static void ListAllCars(ObjectContainer db)
        {
            ObjectSet result = db.Get(typeof(Car));
            ListResult(result);
        }
		// end ListAllCars
        
        public static void StoreCarRollback(ObjectContainer db)
        {
            Pilot pilot = new Pilot("Michael Schumacher", 100);
            Car car = new Car("Ferrari");
            car.Pilot = pilot;
            db.Set(car);
            db.Rollback();
        }
		// end StoreCarRollback
    
        public static void CarSnapshotRollback(ObjectContainer db)
        {
            ObjectSet result = db.Get(new Car("BMW"));
            Car car = (Car)result.Next();
            car.Snapshot();
            db.Set(car);
            db.Rollback();
            Console.WriteLine(car);
        }
		// end CarSnapshotRollback
    
        public static void CarSnapshotRollbackRefresh(ObjectContainer db)
        {
            ObjectSet result=db.Get(new Car("BMW"));
            Car car=(Car)result.Next();
            car.Snapshot();
            db.Set(car);
            db.Rollback();
            db.Ext().Refresh(car, int.MaxValue);
            Console.WriteLine(car);
        }
		// end CarSnapshotRollbackRefresh

		public static void ListResult(ObjectSet result)
		{
			Console.WriteLine(result.Count);
			foreach (object item in result)
			{
				Console.WriteLine(item);
			}
		}
		// end ListResult
    }
}

File: db4on/reference/vb/db4o-reference-chapters/src/com/db4odoc/f1/clientserver/TransactionExample.vb
Imports System
Imports System.IO
Imports com.db4o

Namespace com.db4odoc.f1.clientserver
    Public Class TransactionExample

        Public Shared ReadOnly YapFileName As String = "formula1.yap"

        Public Shared Sub Main(ByVal args As String())
            File.Delete(YapFileName)
            Dim db As ObjectContainer = Db4oFactory.OpenFile(YapFileName)
            Try
                StoreCarCommit(db)
                db.Close()
                db = Db4oFactory.OpenFile(YapFileName)
                ListAllCars(db)
                StoreCarRollback(db)
                db.Close()
                db = Db4oFactory.OpenFile(YapFileName)
                ListAllCars(db)
                CarSnapshotRollback(db)
                CarSnapshotRollbackRefresh(db)
            Finally
                db.Close()
            End Try
        End Sub
        ' end Main

        Public Shared Sub StoreCarCommit(ByVal db As ObjectContainer)
            Dim pilot As Pilot = New Pilot("Rubens Barrichello", 99)
            Dim car As Car = New Car("BMW")
            car.Pilot = pilot
            db.[Set](car)
            db.Commit()
        End Sub
        ' end StoreCarCommit

        Public Shared Sub ListAllCars(ByVal db As ObjectContainer)
            Dim result As ObjectSet = db.[Get](GetType(Car))
            ListResult(result)
        End Sub
        ' end ListAllCars

        Public Shared Sub StoreCarRollback(ByVal db As ObjectContainer)
            Dim pilot As Pilot = New Pilot("Michael Schumacher", 100)
            Dim car As Car = New Car("Ferrari")
            car.Pilot = pilot
            db.[Set](car)
            db.Rollback()
        End Sub
        ' end StoreCarRollback

        Public Shared Sub CarSnapshotRollback(ByVal db As ObjectContainer)
            Dim result As ObjectSet = db.[Get](New Car("BMW"))
            Dim car As Car = DirectCast(result.[Next](), Car)
            car.Snapshot()
            db.[Set](car)
            db.Rollback()
            Console.WriteLine(car)
        End Sub
        ' end CarSnapshotRollback

        Public Shared Sub CarSnapshotRollbackRefresh(ByVal db As ObjectContainer)
            Dim result As ObjectSet = db.[Get](New Car("BMW"))
            Dim car As Car = DirectCast(result.[Next](), Car)
            car.Snapshot()
            db.[Set](car)
            db.Rollback()
            db.Ext().Refresh(car, Integer.MaxValue)
            Console.WriteLine(car)
        End Sub
        ' end CarSnapshotRollbackRefresh

        Public Shared Sub ListResult(ByVal result As ObjectSet)
            Console.WriteLine(result.Count)
            For Each item As Object In result
                Console.WriteLine(item)
            Next
        End Sub
        ' end ListResult
    End Class
End Namespace
